torchvision/datasets/coco.py

```diff
@@ -65,7 +71,11 @@ def __getitem__(self, index):
 
         path = coco.loadImgs(img_id)[0]['file_name']
 
-        img = Image.open(os.path.join(self.root, path)).convert('RGB')
+        if self.root_zip is not None:
+            f = self.root_zip[os.path.split(self.root)[1] + "/" + path]
+            img = Image.open(f).convert('RGB')
+        else:
+            img = Image.open(os.path.join(self.root, path)).convert('RGB')
```

Comment:
I think that we should come up with an abstraction that works in both cases, and we replace `os.path.join` with it.
Or else this will end-up being not scalable, and the code will get more and more complex over time.

Maybe something like

```python
def get_path_or_bytesio(root_or_zip, path):
    if is_dir(root_or_zip):
        return os.path.join(root_or_zip, path)
    with zipfile.ZipFile(root_or_zip, 'r') as f:
        zip_info = f.getinfo(path)
        # rest of the code here
```

Reply:
I've moved part of the logic into VisionDataset. But the hard part is that the logic for mapping file paths onto the zipfile is dataset-specific (see e.g. CelebA I added in the latest version, there the maping is different).

Reply:
From a look at the CelebA changes, it seems to me that it could be made generic with a few changes.

But let's first see how much overhead we have by opening the zip file at every `__getitem__`

Reply:
Made it generic through the get_path_or_fp function.